File: Java/PathSumIII.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public int pathFrom(TreeNode root, int sum) {
        if(root == null) return 0;
        int cnt = 0;
        if(root.val == sum) cnt++;
        int left = pathFrom(root.left, sum - root.val);
        int right = pathFrom(root.right, sum - root.val);

        cnt += (left + right);

        return cnt;
    }

    public int pathSum(TreeNode root, int sum) {
        if(root == null) return 0;

        return pathFrom(root, sum) + pathSum(root.left, sum) + pathSum(root.right, sum);
    }
}

File: Go/path_sum_III.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func pathFrom(root *TreeNode, sum int) int {
    if root == nil {return 0}
    var cnt int
    if root.Val == sum {cnt += 1}

    left := pathFrom(root.Left, sum - root.Val)
    right := pathFrom(root.Right, sum - root.Val)

    return cnt + left + right
}

func pathSum(root *TreeNode, sum int) int {
    if root == nil {return 0}
    return pathFrom(root, sum) + pathSum(root.Left, sum) + pathSum(root.Right, sum)
}